packages/commands/src/index.ts

```diff
@@ -579,6 +583,21 @@ export class CommandRegistry {
     this._startTimer();
   }
 
+  /**
+   * Delay the execution of any command matched against the given 'keydown' event
+   * until the `permission` to execute is granted.
+   *
+   * @param event - The event object for a `'keydown'` event.
+   * @param permission - The promise with value indicating whether to proceed with the execution.
+   *
+   * ### Note
+   * This enables the caller of `processKeydownEvent` to asynchronously prevent the
+   * execution of the command based on external events.
+   */
+  holdKeyBindingExecution(event: KeyboardEvent, permission: Promise<boolean>) {
+    this._holdKeyBindingPromises.set(event, permission);
+  }
```

Comment:
could you elaborate the need for that API? Is it to deal with composed shortcut requiring multiples shortcut to be pressed in sequence?

Reply:
No, this is strictly to enable https://github.com/jupyterlab/jupyterlab/pull/15927.

The intent is to block execution of a shortcut if its keystroke would lead to a character being typed in (as detected when handling `onbeforeinput` event).

The problem is that we do not know if a keypress would result in a character being produced at the time when `keydown` is handled, because only after `keydown` is handled the `onbeforeinput` event is emitted.

Therefore https://github.com/jupyterlab/jupyterlab/pull/15927 passes a promise to `holdKeyBindingExecution` which resolves to `true` if the subsequent `onbeforeinput` event does not produce a character, and `false` if it does.

Reply:
Thanks for the details